### ISSUE TITLE
What does pipenv need to stop vendoring a patched version of pip?

### ISSUE THREAD

**Issue by pradyunsg** [pypa/pipenv#5379]
The question says it all really. I'd like for the Python packaging ecosystem to move away from trying to use pip as a library, to broadly reduce the fragility to the tooling.

What are the parts of pip that pipenv tries to use, and what would be necessary to decouple pipenv from pip, so that it can use a shared implementation (eg: build, installer) instead?

**Comment by pradyunsg** [comment]
Adding a cross reference to https://github.com/pypa/pip/pull/10964

**Comment by matteius** [comment]
@pradyunsg I'll eventually have more time to elaborate, but my initial take is this would be ultimately very challenging for both parties, as we use many internal aspects of pip across pipenv and requirementslib.   We vendor pip to ensure the internals are reliable for what is integrated against, and there are even a few patch file we apply:  https://github.com/pypa/pipenv/tree/main/tasks/vendoring/patches/patched

A big part of the internals we use is the resolver, for which there is no public interface for.   We also solved for package confusion attacks with the index restriction patch:  https://github.com/pypa/pipenv/blob/main/tasks/vendoring/patches/patched/pip_index_safety.patch
I opened a PR against pip for this but it was the same problem of using pip internals so not gaining acceptance:  https://github.com/pypa/pip/pull/10964

We also have reduced the scope of our other vendor'd packages to reuse other vendor'd libraries provided by pip.   Here are some imports we are currently referencing:
![image](https://user-images.githubusercontent.com/479892/193281011-4f2c4886-cd44-4a01-93fe-7bd1d3edee71.png)
![image](https://user-images.githubusercontent.com/479892/193281198-8c0152e4-5bfe-404c-acf7-548a91b8e101.png)

Plus a ton in requirementslib which we vendor in:
![image](https://user-images.githubusercontent.com/479892/193281378-982fe00a-6a3d-46de-b97a-378903afc601.png)
![image](https://user-images.githubusercontent.com/479892/193281473-89f5cbef-38ba-4684-bc1f-82126c49ee06.png)

**Comment by pradyunsg** [comment]
> Why is it ok for pip to do this but advocate for pipenv to not? (I get the argument that you don't view pip as a library)

None of the arguments other than fragility from https://pip.pypa.io/en/stable/development/vendoring-policy/#rationale apply.

Having pipenv live outside of the environments it manages and outside of a user-managed environment where they could run a command and break it, would solve the fragility problem while avoiding all the caveats that come with vendoring.

**Comment by pradyunsg** [comment]
I'll also note that currently, pipenv isn't merely vendoring dependencies but vendoring forks of those dependencies, given that you have patches to implement entire features in the packages/dependencies that you vendor.

That means that it's harder for downstream redistributors to redistribute the package and harder for you to keep up to date with the dependency (you are effectively managing a fork at that point). That's... not particularly sustainable IMO.

**Comment by matteius** [comment]
> This sort of shit is exactly _why_ pip isn't supposed to be used as a library in fact. :P

But that is actually an example where we use sub-process and invoke pip "normally" using the public interfaces :-D

Thanks again -- I've been pretty busy with a construction project and being off work for two weeks, this week I've been back to work so its been a real whirlwind.   

From the perspective of what could pip do to improve things, If there were a public interface to the resolver that we could use for the lock phase, that would be something we would try to integrate with over using the internal APIs to the resolver.   Maybe this interface could make use of similar code to that reference PR from the index restricted packages, and then there would be a use case for that on both sides.  We could pick apart the interface we currently have into the resolver to try and negotiate what the public interface could be like.   This should be a lot more feasible to understand the current code now that `pip-shims` is not involved in obfuscating everything.

**Comment by oz123** [comment]
@pradyunsg thank you for your interest in pipenv development.
Matt has already answered many of your concerns. I'll just add my two cents:
First,
>  well, this project has a huge pile of technical debt from a series of really bad design decisions early on in its lifecycle. :)

Can't agree more here with you. Pipenv was and is very popular, and it's not bad we have also other solutions, that compete in this space. However, not all of them do all that it does, and sometimes it's hard for users to move away to a new project.
So, here we are humbly working almost for free to fix this situation. Pipenv was really in a bad shape but we are trying to keep the user interface and really rewriting it from the inside.
Matt has done a tremendous job here on requirementslib and the resolver.
I was more busy cleaning those forks and all those vendored libraries we were dragging.

> FWIW, it sounds like pipenv should invest in setting up an installer

Actually, we do have an installer, which is based on the pip installer. 

https://github.com/pypa/pipenv/blob/main/get-pipenv.py

However, you can also just use pip. 

We'd like to see all the issues you mentioned fixed, but being just two devs with full time jobs, our time is limited. Hence, this will take time to fix everything. However, rest assured they will be fixed.

**Comment by oz123** [comment]
BTW, I think we should move this to a discussion instead of an issued ...

**Comment by pradyunsg** [comment]
> we are trying to keep the user interface and really rewriting it from the inside.

Same, with pip. :)

> Actually, we do have an installer, which is based on the pip installer.
> 
> [`main`/get-pipenv.py](https://github.com/pypa/pipenv/blob/main/get-pipenv.py)

That's not what I meant by "installer" -- Poetry's installer installs poetry in a separate virtualenv, placed at a well-known location *outside* of the user's Python environments.

That way, Poetry lives with its own environment (like an application). This environment cannot be broken by changing packages via `pip` unless the user goes out of their way to do that.

**Comment by frostming** [comment]
> I'll note that none of this is designed to be a library. This sort of shit is exactly _why_ pip isn't supposed to be used as a library in fact. :P

True, but there is a growing need to reuse some part of pip internals, and few has alternatives:

- resolver - [resolvelib] but it is a rather *abstract* framework and needs concrete implementations.
- requirement parsing - [requirementslib] which also utilizes `pip`'s internals.
- package finder - I made [one][unearth].
- installer - there are [distlib] and [installer], but they will likewise fail on some corner cases that pip has put big efforts into supporting them.
- There are also other unofficial libraries like [pip-api] to cover a limited subset of `pip`'s internals.

[resolvelib]: https://pypi.org/project/resolvelib
[requirementslib]: https://pypi.org/project/requirementslib
[unearth]: https://pypi.org/project/unearth
[pip-api]: https://pypi.org/project/pip-api
[distlib]: https://pypi.org/project/distlib
[installer]: https://pypi.org/project/installer

Most importantly, none of the above alternatives is as battle-tested as pip and the open-source maintainers have to work hard on keeping the behaviors consistent with pip. So it is a reasonable choice to vendor the entire pip as a library for pipenv.

IMO the situation won't improve until pip either extract all the components into independent libraries, or replace the internal functionalities with some of the open-source alternatives.

**Comment by oz123** [comment]
If someone uses pip as a library and it's working for him, I don't see why pip developmer should care. Ofcourse if this same people always complained about it in the pip channels, the response could be: we told you not to do so.
On the other hand, if some people use it and there is a need for it, pip development community could yield and make a public API (which isn't through shell commands).
Ofcourse, these are just my two cents. However, the fact is that there are many python package managers and virtual environment managers ( pipenv, conda, hatch, poetry  has all existed for a long long time) shows that there is a long time dissatisfaction with python package management. 
The maintenanrs of this project (myself including) scratch their own itch or have strong interest in package management, and thus decided to put effort into solving the problem the way they see fit. I honestly don't see why that interferes with pip development.

**Comment by oz123** [comment]
If pip offered a public API, we could stop vendoring it.
Here is another attempt to make a public API for pip.

https://github.com/reynoldsnlp/pipster

**Comment by kalebmckale** [comment]
My suggestion for isolating `pipenv`, or at least what I do, is install it via `pipx`. No reason to reinvent the wheel when a solution exists. Perhaps we could document this as an option? Now, I'm second-guessing if there is already reference to this somewhere in the documentation. 😁

**Comment by frostming** [comment]
> My suggestion for isolating `pipenv`, or at least what I do, is install it via `pipx`

True, `pipx` isolates things quite well, but there are still people who installs the tool under the global site-packages. It may cause conflicts even in an isolated venv, since `pip` is such a fundamental library that exists in every venv. IIRC pipx installs pip in a shared location and inject the path via `pth` file. I suspect it may conflict with the pipenv dependency.

**Comment by kalebmckale** [comment]
No, no, I'm not suggesting to use `pipx` to isolate `pip` but `pipenv`. Different tools that create virtual environments often include their own copies of `pip` anyhow.

In practice, my user/local site-packages directory is bare-to-empty. I have a number of tools installed via `pipx`, including `pipenv` and inject related libraries and plugins for those application environments. Any packages related to a project or needed for development or testing of the project are included in categories of `Pipfile` (to ensure no dependency conflicts for users and developers alike) and installed (as needed) by `pipenv` into the project's virtual environment, created by `pipenv`.

Anyhoo... some of that may be a bit off-topic but just thought I'd share how my workflow uses `pipenv`.

**Comment by oz123** [comment]
@matteius sorry for necroing this issue, but it seems that Dusting is still maintaining pip-api, maybe we should join forces and start using pip-api, instead of vendoring pip? Will it make it easier for us to maintain pipenv?
